File: apps/desktop/src/data/library/items.ts
import { convertFileSrc, invoke } from "@tauri-apps/api/core";
import type * as LibraryContract from "../../contracts";

const command = {
  create: "create_library_item",
  page: "get_library_items_page",
  update: "update_library_item",
  delete: "delete_library_item",
  cancel: "cancel_library_transcription",
  retry: "retry_library_transcription",
  export: "export_library_item_to_path",
  tags: "get_library_tags",
  translations: "get_library_translations",
  translate: "translate_library_item",
  deleteTranslation: "delete_library_translation",
} as const;

type NativePayload = Record<string, unknown>;

const run = <Result>(name: string, payload?: NativePayload) =>
  payload === undefined ? invoke<Result>(name) : invoke<Result>(name, payload);

const runForItem = (name: string, id: string) => run<void>(name, { id });

export const resolveLibraryAudioUrl = (audioPath: string): string =>
  convertFileSrc(audioPath);

type CreateArguments = [
  path: string,
  options: LibraryContract.LibraryImportOptions,
];
export const createLibraryItem = (...[path, options]: CreateArguments) =>
  run<LibraryContract.LibraryItem>(command.create, { path, options });

type PageArguments = [
  filter: LibraryContract.LibraryFilter,
  limit: number,
  offset: number,
];
export const getLibraryItemsPage = (
  ...[filter, limit, offset]: PageArguments
) =>
  run<LibraryContract.LibraryItemsPage>(command.page, {
    filter,
    limit,
    offset,
  });

export const updateLibraryItem = (
  id: string,
  patch: LibraryContract.LibraryItemPatch,
) => run<LibraryContract.LibraryItem>(command.update, { id, patch });

export const deleteLibraryItem = (id: string) => runForItem(command.delete, id);
export const cancelLibraryTranscription = (id: string) =>
  runForItem(command.cancel, id);
export const retryLibraryTranscription = (id: string) =>
  runForItem(command.retry, id);

type ExportArguments = [
  id: string,
  format: LibraryContract.ExportFormat,
  outputPath: string,
];
export const exportLibraryItemToPath = (
  ...[id, format, outputPath]: ExportArguments
) => run<void>(command.export, { id, format, outputPath });

export const getLibraryTags = () => run<string[]>(command.tags);

export const getLibraryTranslations = (itemId: string) =>
  run<LibraryContract.LibraryTranslation[]>(command.translations, { itemId });

export const translateLibraryItem = (itemId: string, language: string) =>
  run<LibraryContract.LibraryTranslation>(command.translate, {
    itemId,
    language,
  });

export const deleteLibraryTranslation = (itemId: string, language: string) =>
  run<void>(command.deleteTranslation, { itemId, language });

File: apps/desktop/src-tauri/src/library/commands/items.rs
use tauri::AppHandle;

use crate::{AppRuntime, AppState, LibraryJob};

use super::super::{
    processing::{build_export_content, build_meeting_export_content},
    queue::{release_library_slot, schedule_library_job},
    types::{
        ExportFormat, LibraryFilter, LibraryItem, LibraryItemPatch, LibraryItemStatus,
        LibraryItemsPage,
    },
};
use super::{files, recovery};

pub(super) fn page(
    state: &AppState,
    filter: Option<LibraryFilter>,
    limit: u32,
    offset: u32,
) -> Result<LibraryItemsPage, String> {
    let page_size = limit.clamp(1, 200) as usize;
    state
        .storage()
        .get_library_items_page(filter.unwrap_or_default(), page_size, offset as usize)
        .map(|(items, has_more)| LibraryItemsPage { items, has_more })
        .map_err(|error| format!("Failed to load library items page: {error}"))
}

pub(super) fn update(
    state: &AppState,
    id: String,
    patch: LibraryItemPatch,
) -> Result<LibraryItem, String> {
    let storage = state.storage();
    let item = storage
        .update_library_item(&id, patch)
        .map_err(|error| format!("Failed to update library item: {error}"))?
        .ok_or_else(|| "Library item not found".to_owned())?;
    let mirror_settings = state.current_settings_unmasked();
    if let Err(error) =
        crate::markdown_mirror::mirror_library_by_id(&mirror_settings, &storage, &id)
    {
        tracing::warn!("Failed to update Markdown mirror for Library item: {error}");
    }
    Ok(item)
}

pub(super) fn delete(
    app: &AppHandle<AppRuntime>,
    state: &AppState,
    id: String,
) -> Result<(), String> {
    if state.meeting_capture().is_active()
        && state.meeting_capture().state().id.as_deref() == Some(id.as_str())
    {
        return Err("Stop the meeting recording before deleting it.".to_owned());
    }

    state.remove_library_job(&id);
    state.cancel_library_transcription(&id);
    release_library_slot(app, state, &id);

    let storage = state.storage();
    let Some(item) = storage
        .get_library_item(&id)
        .map_err(|error| format!("Failed to load library item: {error}"))?
    else {
        return Ok(());
    };
    files::delete_managed_audio(app, &item.audio_path)?;
    storage
        .delete_library_item(&id)
        .map(|_| ())
        .map_err(|error| format!("Failed to delete library item: {error}"))
}

pub(super) fn cancel(app: &AppHandle<AppRuntime>, state: &AppState, id: String) {
    if state.remove_library_job(&id) {
        recovery::update_status(&state.storage(), &id, LibraryItemStatus::Cancelled);
        recovery::emit_error(app, &id, "Transcription cancelled", true);
        return;
    }
    state.cancel_library_transcription(&id);
    recovery::update_status(&state.storage(), &id, LibraryItemStatus::Cancelling);
}

pub(super) fn retry(
    app: &AppHandle<AppRuntime>,
    state: &AppState,
    id: String,
) -> Result<(), String> {
    let storage = state.storage();
    let item = load_item(state, &id)?;
    let kind = match recovery::retry_job(&item) {
        Ok(kind) => kind,
        Err(message) => {
            recovery::set_error(&storage, &id, &message);
            recovery::emit_error(app, &id, message.clone(), false);
            return Err(message);
        }
    };
    recovery::update_status(&storage, &id, LibraryItemStatus::Pending);
    schedule_library_job(app, state, LibraryJob { id, kind });
    Ok(())
}

pub(super) fn export(
    state: &AppState,
    id: String,
    format: ExportFormat,
    output_path: String,
) -> Result<(), String> {
    let item = load_item(state, &id)?;
    let content = if item.kind == "meeting" && matches!(&format, ExportFormat::Md) {
        let details = state
            .storage()
            .get_meeting_details(&id)
            .map_err(|error| format!("Failed to load meeting details: {error}"))?
            .ok_or_else(|| "Meeting details not found".to_owned())?;
        build_meeting_export_content(&item, &details, format)
    } else {
        build_export_content(&item, format)
    }
    .map_err(|error| format!("Failed to build export: {error}"))?;
    files::write_export(&output_path, &content)
}

pub(super) fn tags(state: &AppState) -> Result<Vec<String>, String> {
    state
        .storage()
        .get_library_tags()
        .map_err(|error| format!("Failed to load tags: {error}"))
}

fn load_item(state: &AppState, id: &str) -> Result<LibraryItem, String> {
    state
        .storage()
        .get_library_item(id)
        .map_err(|error| format!("Failed to load library item: {error}"))?
        .ok_or_else(|| "Library item not found".to_owned())
}
